docs/design/README.md

```diff
@@ -534,6 +534,149 @@ After installing packages, the result of `dnf list installed` will be output to
+
+When creating a kit, we won't have the yum install command to help us figure out all the right packages.
+We need to change the build system so that the dependency graph can be used to put the right packages into a kit.
+
+For starters, we will no longer rely on a flat structure in `build/rpms`.
+We will put the output of each `package` build into a subdirectory using the Cargo package's name.
+
+For example, when building `glibc/Cargo.toml` we will put the output in `build/rpms/glibc`.
+Note that more than one RPM file can result from a package build.
+The subdirectory maps the resultant RPM files to the Cargo package that produced them.
+
+At each `build-package` invocation, a distinct, flat RPM directory will be created containing the packages needed for the current `build-package` invocation.
+This will be done by traversing the Cargo dependency graph and making a list of packages and kits that are needed.
+The kits will be unpacked and the direct package dependencies will be copied from their `build/rpms` subdirectories.
+This will require us to pass a list of the packages needed for the build to the dockerfile, for example:
+
+```
+PACKAGES=pkg-a pkg-b pkg-c
+```
+
+This list needs to be complete and include only "pure" package dependencies.
+That is, it should not include any package names that are "brought in" by a kit dependency.
+The packages found in each of those subdirectories, e.g. `build/rpms/pkg-a`, will be copied to a flat RPM directory for the build.
+
+Additionally, the `build-package` command needs to know the list of kits required:
+
+```
+KITS=core-kit aws-kit
+```
+
+Each of these needs to be unpacked and have its RPMs copied out to the flat RPM directory.
+For example, these would be unpacked: `build/kits/core-kit.tar`, `build/kits/aws-kit.tar`.
+
+The `build-kit` requirements are slightly simplified.
+Only the list of "pure" package dependencies is required.
+When constructing a kit, buildsys does not need to know what other kits the current kit depends on.
+
+#### Dependency Traversal
+
+The simplest way to get a list of "pure" package dependencies and a list of kits is to traverse the dependency graph twice.
+
+In the first traversal, kit dependencies are ignored and not followed, resulting in only the "pure" package dependencies.
+In the second traversal, all nodes are followed and package dependencies are filtered out from the final list, resulting in a list of all required kits.
+
+#### Pseudo Code
+
+**Package Build Time**
+
+buildsys:
+
+```
+// load the package's manifest
+let (packages, kits) = manifest.walk_dependency_graph();
+
+// when calling build_package:
+  build-arg PACKAGES="{packages}"
+  build-arg KITS="{kits}"
+```
+
+Dockerfile:
+
+```
+ARG PACKAGES
+ARG KITS
+
+mkdir big-rpm-dir
+
+# PACKAGES could be empty for packages like glibc with no deps
+for pkg in ${PACKAGES} ; do
+  cp /host/build/rpms/${pkg}/* big-rpm-dir
```

Comment:
This syncs up with above question on why the pkg dirs are needed especially if the end result is just copying them to the kit anyway. Seems like creating duplicate storage in build folder for an unknown reason right now